### ISSUE TITLE
Create Code Climate engine

### ISSUE THREAD

**Issue by ai** [stylelint/stylelint#1341]
If Code Climate will move to Stylelint it will be awesome for promotion.

I think @davidtheclark should write a mail to Code Climate team, explain benefits of Stylelint (compare to csslint) and offer any help :) (I will help you they will ask something complicated).

**Comment by jeddy3** [comment]
They are looking for a community contribution, via the [developer program](https://codeclimate.com/dev_form). 

Ref: https://twitter.com/a_baldwinhunter/status/733747324267167749

**Comment by davidtheclark** [comment]
I guess at some point someone has to write that integration for them ...

**Comment by jeddy3** [comment]
I'm hoping that that someone is a person or team who uses both Code Climate and stylelint, and has a need to fill the hole :) 

Labelled as "help wanted" accordingly.

**Comment by ntwb** [comment]
Via https://github.com/codeclimate/codeclimate-eslint/issues/80#issuecomment-223386512

> > ntwb wrote: 
> > "Over here stylelint/stylelint#1341 we have been thinking/wanting to write a stylelint engine for Code Climate, maybe not replace the existing CSSLint CSS Code Climate engine, but to start with as an alternative for sure 😏"

> Code Climate wrote:
> "That's fantastic! We've chatted about stylelint here at Code Climate HQ and it would be great to have an engine available. stylelint would make a great native engine! Let me @dblandin or @mrb know if you have any more questions regarding the engine submission process."
>
> "We have a codeclimate-community slack organization where members hang out to discuss engine ideas and development. If you're interested, you can join our Developer Program to get an invite."
 

👍 
***

A slight caveat on the use of stylelint shared plugins and configs with Code Climate:

>  > > Code Climate Wrote:
> > > "Unlike many popular CI systems, our build infrastructure isn't yet designed to execute customer code. It was designed to run engines statically against source code. Before we allow arbitrary execution of customer code (including NPM packages or RubyGems), we have some work to do to ensure that our systems and customer data remain secure. Unfortunately, this isn't an issue specific to ESLint. It's a shared concern across our platform."

> > ntwb wrote:
> > "If we were to write a stylelint Code Climate engine stylelint would also be affected by the above in that Code Climate wouldn't allow us to load stylelint shared plugins or shared configs to be used with the stylelint engine correct?"

> Code Climate wrote:
> "That's true. A stylelint Code Climate engine would also be affected by the same issue in that stylelint plugins or shared configs that aren't explicitly vendored within the engine itself would be unavailable during analysis on codeclimate.com."
>
> "We would be happy to work with you as an engine author to figure out which plugins/styles should be vendored into a stylelint engine and available to any Code Climate user either via the CLI or on codeclimate.com."

**Comment by davidtheclark** [comment]
Thanks for filling us in, @ntwb.

**Comment by jeddy3** [comment]
Agreed. Thanks for digging into this.

> That's true. A stylelint Code Climate engine would also be affected by the same issue in that stylelint plugins or shared configs that aren't explicitly vendored within the engine itself would be unavailable during analysis on codeclimate.com."

Unfortunately, does that mean stylelint support is moot until their systems allow "arbitrary execution of customer code"?

**Comment by ntwb** [comment]
> Unfortunately, does that mean stylelint support is moot until their systems allow "arbitrary execution of customer code"?

Pretty much I believe this to be the case, the original code climate issue linked above was a request to whitelist our `eslint-config-stylelint` shared ESLint configuration, this request was denied:
> "We've made a few exceptions for popular plugins that we now include within the engine."

If we were to create a stylelint code climate engine we'd probably be able to get `stylelint-config-standard` included and maybe a couple of others in the stylelint engine, any other shared configs, or plugins would presumably have to be approved by the stylelint team (possibly also code climate) and then manually added to the stylelint code climate engine. None of the maintenance of the shared plugins, shared configs, or the stylelint engine itself would be trivial from having _watched_ the Code Climate ESLint repo for approximately the past 3 months.

Code Climate customers/users whom want to use their own stylelint config would have to include the entire config code in their repo. A valid cosmiconfig configuration I believe wouldn't be an issue, the key here is that the entire config be in the repo, or if it extends an existing config, e.g. `stylelint-confg-standard` unless that shared config is included in the stylelint engine then you're out of luck.

I'm tempted to close this issue as `maybe later` for when Code Climate do support extending and sharing plugins and configs in the stylelint (and ESLint) ecosystem that work so well. I doubt the work involved in initially creating a code climate engine is not insignificant (nor the maintenance outlined above) and at this current point in time once created would only support a fraction of stylelints users.

**Comment by jeddy3** [comment]
Thanks for writing this up!

> I doubt the work involved in initially creating a code climate engine is not insignificant (nor the maintenance outlined above) and at this current point in time once created would only support a fraction of stylelints users.

I agree with this assessment.

> I'm tempted to close this issue as maybe later

Lets keep this issue open for now in case someone comes along willing to champion this integrating and to put in the legwork.

**Comment by gilbarbara** [comment]
Hey,
I just started build an engine for codeclimate. I hope to have it ready by next week.
https://github.com/gilbarbara/codeclimate-stylelint

**Comment by evilebottnawi** [comment]
@gilbarbara Very thanks!

**Comment by evilebottnawi** [comment]
/cc @davidtheclark @jeddy3 @gilbarbara 
I propose to transfer the project to our organization. We will be able to correct mistakes and develop if there was free time.

**Comment by jeddy3** [comment]
> I propose to transfer the project to our organization.

I think pushing the 3rd party configs out of the org and into the wider community worked out well for us. It helped us establish a _clear and definable line_ about what belongs in the org (and is our responsibility) and what does not. Therefore, I'm in favour of _not_ transferring this project to the org as it's not part of the core system.

> We will be able to correct mistakes and develop if there was free time.

You can fork the project for that.

**Comment by davidtheclark** [comment]
Agree with @jeddy3 ^^. Why do you think this particular project is different from others, like Atom linter, etc., @evilebottnawi? We can still help out @gilbarbara if needed.

**Comment by evilebottnawi** [comment]
@davidtheclark @jeddy3 you convinced me :smile:

**Comment by gilbarbara** [comment]
Hello!
The engine is in beta! You can enable it in your `.codeclimate.yml` with:
```yaml
engines:
  stylelint:
    enabled: true
    channel: "beta"
```

It will run on codeclimate.com or with the CLI.

And there's already a new version with extended rule info info that would probably go live early next week.

**Comment by gilbarbara** [comment]
Ahh,

Since codeclimate don't install the node_modules while testing I had to add support in the engine for configs and plugins. I've enable stylelint-config-standard and stylelint-scss..
If you think it should support other packages let me know.

**Comment by jeddy3** [comment]
@gilbarbara Excellent job. Thank you!

> If you think it should support other packages let me know.

We have a _curated_ [list of plugins](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/plugins.md). Perhaps all of them should supported?

Having said that, the list is in [flux](https://github.com/stylelint/stylelint/issues/2010) at the moment and so give me a mo to fix it up before you consider it.

**Comment by gilbarbara** [comment]
Hey @jeddy3 
Thanks! 😁 

Let me know when you're done with the list.

Also, codeclimate has a "soft" requirement to add a readup to the issues it reports. A short description of the rule and a simple example if possible.
So I ended up cloning the styleslint repo (stylelint-scss too), parsing the `rules` directory, extracting an excerpt from the README files and saving it as json..
However some of the repos on the list of plugins don't have a separated file for its rules. If the list stays small I could add those description manually but if you have an idea how to automate this process, let me know.

[s]

**Comment by jeddy3** [comment]
> Let me know when you're done with the list.

It's [done](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/plugins.md). Sorry, I forgot to mention it last week.

> If the list stays small

We can't guarantee that it will stay small I'm afraid. Although it has pretty much stayed 6 or so items for the last few month. Perhaps using that list isn't the best approach in the long run, although it works for now and I can't think of any alternatives. 

> So I ended up cloning the styleslint repo (stylelint-scss too), parsing the rules directory, extracting an excerpt from the README files and saving it as json..

Sounds like the best approach given the requirements!

> if you have an idea how to automate this process, let me know.

I'm afraid not. I doubt they'll be many more plugin _packs_ in the near future and so we can always revisit this later. I think relying on PRs to your repo is fine for now for new plugins.

I'm going to create a PR that adds your engine to the docs here. That'll point people in the right direction should they wish to contribute to the engine. We can close this issue once that PR is merged.

**Comment by gilbarbara** [comment]
@jeddy3 Great.
I think the engine will enter the `stable` status next week or so.

Support for JetBrains IDEs and codeclimate, I've been waiting this for quite sometime. Bye scss-lint! 😁

**Comment by jeddy3** [comment]
> I think the engine will enter the stable status next week or so.

Excellent. Thanks again for your work on this!

**Comment by gilbarbara** [comment]
@jeddy3 
The engine reached `stable` and is featured on codeclimate's engine page. 🔥

**Comment by jeddy3** [comment]
Awesome thanks!
